### PR TITLE
Grading Improvements 

### PR DESCRIPTION
**Pull request by geshuming**
AutogradingTabResponsive

- Main features: Improved styling for testcases

- Added styling to Autograding Tab. Students and avengers are now able to see whether they pass or fail test cases.

- Added ResultCard, students and avengers are now able to see autograding results.

- Added titles to Editors

- Added Timeout Error Explanation

- Improved overall CSS.

- Slight cleanup

- Fixed issue #570

- Updated frontend mocks for 1920 Source Format

### CODE REVIEW

**Comment by geshuming** [comment]
It is just improved styling over the testcases that we already have. Users will only need to press 'test' to run the corresponding testcase against their code. The expected and actual outcome are compared and there will be an effect to show if a user has passed or failed the testcase.

There is also a section below that displays results from the autograder, after a student has submitted.

**Comment by geshuming** [comment]
Yep we are migrating to [a new xml format for missions to ease the autograding process](https://github.com/source-academy/assessments/wiki/Assessments). 

1) As a result, it won't be backwards compatible for the autograder only and the missions will have to be updated.

2 & 3) I agree. I think we can simplify the UI/UX elements too. Maybe we can clean this up later? 

4) Prepend code is part of the xml.

5) I think we were in favor of giving sample tests for students to run, but also allow room for students to explore edge cases. I hope that the use of providing testcases will help prevent cluttering the student's editor with a lot of code, since past assessments had missions with various "testcases" inside the code editor itself, and students had to comment/uncomment the function calls to test their code (which is not ideal in my opinion)

@martin-henz do you have any other views?

**Comment by rrtheonlyone** [comment]
Just checking: Right now, in our current `master` branch, does this feature exist?

**Comment by geshuming** [comment]
Yes. This PR is meant to extend the previous grading features.

**Comment by martin-henz** [comment]
About:

**@martin-henz do you have any other views?**

Wearing the overall management hat: This feature is on the critical path, right now. We need to finalize the mission format and start migrating our content.

So I suggest that we get the functionality right and a working interface. Improvements in the interface can then be handled with an Issue.

Thanks.

File: src/components/workspace/index.tsx [hunk 1]
```diff
@@ -204,13 +204,15 @@ class Workspace extends React.Component<WorkspaceProps, {}> {
       ) {
         return (
           <div className="editor-content">
+            <div className="editor-title">Prepend Code</div>
```
[the remark this answers is not in the document]
Comment (rrtheonlyone) [reply src/components/workspace/index.tsx]: Plus what does "prepend" even mean? Can we go with a more non-technical term that a student can associate with => such as "available functions"

File: src/components/workspace/side-content/AutograderCard.tsx [hunk 2]
```diff
@@ -47,45 +24,40 @@ class AutograderCard extends React.Component<AutograderCardProps, {}> {
       }
     };
 
+    const isCorrect =
+      this.props.testcase.actual !== undefined
+        ? // tslint:disable-next-line
```
Comment (rrtheonlyone) [inline src/components/workspace/side-content/AutograderCard.tsx]: why is tslint disabled here?

File: src/components/workspace/side-content/ResultCard.tsx [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+              <div className="row error-explanation">
+                {'Error: '}
+                <pre className="code">{'Line ' + error.line + ': ' + error.errorExplanation}</pre>
+              </div>
+            </div>
+          );
+      }
+    };
+
+    // tslint:disable-next-line
```
Comment (rrtheonlyone) [inline src/components/workspace/side-content/ResultCard.tsx]: Same here

File: src/mocks/assessmentAPI.ts [hunk 4]
```diff
@@ -208,10 +208,33 @@ What's your favourite dinner food?
...
-    solutionTemplate: '0th question mock solution template',
+    prepend: `// This is a mock Prepend!
+function fibonacci(n) {
+  if (n <= 2) {
+    return 1;
+  } else {
+    return fibonacci(n-1) + fibonacci(n-2);
+  }
+}`,
+    postpend: "// This is a mock Postpend! You shouldn't be able to see me!",
```
Comment (rrtheonlyone) [inline src/mocks/assessmentAPI.ts]: If we have postpend and prepend, why are we only showing prepend on the UI? Or is postpend being shown also?

File: src/reducers/workspaces.ts [hunk 5]
```diff
@@ -359,6 +368,21 @@ export const reducer: Reducer<IWorkspaceManagerState> = (
           isRunning: false
         }
       };
+    case EVAL_TESTCASE_FAILURE:
+      return {
+        ...state,
+        [location]: {
+          ...state[location],
+          editorTestcases: state[location].editorTestcases.map((testcase, i) => {
+            if (i === index) {
```
Comment (rrtheonlyone) [inline src/reducers/workspaces.ts]: Why is `index` retrieved at the very top when your only using it for two cases? Why not get it from the `index` explicitly in the case itself so its clear where it is coming from? (its easier to see where it is coming from)

File: src/reducers/workspaces.ts [hunk 6]
```diff
@@ -301,7 +302,15 @@ export const reducer: Reducer<IWorkspaceManagerState> = (
         ...state,
         [location]: {
           ...state[location],
-          isRunning: true
+          isRunning: true,
+          editorTestcases: state[location].editorTestcases.map((testcase, i) => {
+            if (i === action.payload.testcaseId) {
+              testcase.actual = undefined;
```
Comment (rrtheonlyone) [inline src/reducers/workspaces.ts]: what is `actual` supposed to be? It is hard to tell from the name plus it seems to have `any` type?

File: src/components/workspace/index.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (geshuming) [reply src/components/workspace/index.tsx]: I removed the title for the editors

[the remark this answers is not in the document]
Comment (rrtheonlyone) [reply src/components/workspace/index.tsx]: Removing the title will make it hard to ascertain why the prepend code editor is even there in the first place. 

Possible to move the prepend code portion out to one of the tabs/hide it on a button press? 
Or if thats too difficult, why not call the prepend code, "Available functions" ?

[the remark this answers is not in the document]
Comment (geshuming) [reply src/components/workspace/index.tsx]: The prepend is resizable so it can be hidden by resizing it.

The prepend can contain any code, such as constants and functions, so I don't think calling it "available functions" is ok, maybe "Helpers"? I think we can use a `<Collapse>` to tag the editor as well as introduce a mechanism to hide the prepend editor.

File: src/components/workspace/side-content/AutograderCard.tsx (hunk 2 above)
Comment (geshuming) [reply src/components/workspace/side-content/AutograderCard.tsx]: Refactored the code. I had an issue with === equality returning false.

File: src/components/workspace/side-content/ResultCard.tsx (hunk 3 above)
Comment (geshuming) [reply src/components/workspace/side-content/ResultCard.tsx]: Removed

File: src/mocks/assessmentAPI.ts (hunk 4 above)
Comment (geshuming) [reply src/mocks/assessmentAPI.ts]: The postpend is to allow mission authors to set up helper functions for grading purposes, so it is not displayed for users, but needed for grading purposes.

File: src/reducers/workspaces.ts (hunk 5 above)
Comment (geshuming) [reply src/reducers/workspaces.ts]: Refactored to remove the constant `index`

File: src/reducers/workspaces.ts (hunk 6 above)
Comment (geshuming) [reply src/reducers/workspaces.ts]: Refactored

File: src/components/workspace/index.tsx (hunk 1 above)
[the remark this answers is not in the document]
Comment (rrtheonlyone) [reply src/components/workspace/index.tsx]: Okay that sounds good

File: src/mocks/assessmentAPI.ts (hunk 4 above)
Comment (rrtheonlyone) [reply src/mocks/assessmentAPI.ts]: Okay understood 👍

File: src/components/academy/grading/GradingWorkspace.tsx [hunk 7]
```diff
@@ -180,33 +194,38 @@ class GradingWorkspace extends React.Component<GradingWorkspaceProps> {
       this.props.storedQuestionId !== questionId
```
Comment (rrtheonlyone) [inline src/components/academy/grading/GradingWorkspace.tsx]: Again for this:

```
if (this.props.storedSubmissionId === submissionId && this.props.storedQuestionId === questionId) {
      return;
}
```

File: src/components/academy/grading/GradingWorkspace.tsx [hunk 8]
```diff
@@ -180,33 +194,38 @@ class GradingWorkspace extends React.Component<GradingWorkspaceProps> {
...
     ) {
       const question = this.props.grading[questionId].question as IQuestion;
+      const autogradingResults =
```
Comment (rrtheonlyone) [inline src/components/academy/grading/GradingWorkspace.tsx]: Ternary operators are to be only used for really obvious cases. It hinders readability and its bad practice in general (esp stacking them like how you have done here!)

Lets make things explicit:
```
    let autogradingResults : AutogradingResult[] | undefined;
    let editorValue: string | undefined = ""; 
    let editorPrepend: string | null = "";
    let editorPostpend: string | null = "";
    let editorTestcases: ITestcase[] = [];

    if (question.type === QuestionTypes.programming) {
      autogradingResults = (question as IAnsweredQuestion).autogradingResults;

      editorValue = ((question as IAnsweredQuestion).answer as string);
      if (!editorValue) {
          editorValue = (question as IAnsweredQuestion).solutionTemplate;
      }

      if ((question as IAnsweredQuestion).prepend) {
        editorPrepend = (question as IAnsweredQuestion).prepend;
      }

      if ((question as IAnsweredQuestion).postpend) {
        editorPostpend = (question as IAnsweredQuestion).postpend;
      }

      if ((question as IAnsweredQuestion).testcases) {
        editorTestcases = (question as IAnsweredQuestion).testcases;
      }
    }
```

File: src/components/assessment/assessmentShape.ts [hunk 9]
```diff
@@ -81,7 +82,7 @@ export interface ITestcase {
   answer: string;
   score: number;
   program: string;
-  actual?: any;
+  actual?: string;
```
Comment (rrtheonlyone) [inline src/components/assessment/assessmentShape.ts]: rename this to result?

File: src/components/assessment/assessmentShape.ts [hunk 10]
```diff
@@ -149,3 +150,18 @@ export type Library = {
     2?: string; // For mission control
   }>;
 };
+
+export type AutogradingResult = {
+  resultType: string;
+  expected?: string;
+  actual?: string;
```
Comment (rrtheonlyone) [inline src/components/assessment/assessmentShape.ts]: same for this

File: src/components/workspace/side-content/AutograderCard.tsx [hunk 11]
```diff
@@ -47,45 +24,39 @@ class AutograderCard extends React.Component<AutograderCardProps, {}> {
       }
     };
 
+    const isCorrect =
```
Comment (rrtheonlyone) [inline src/components/workspace/side-content/AutograderCard.tsx]: Please rewrite this ternary portion and make things more readable.

File: src/components/assessment/assessmentShape.ts (hunk 9 above)
Comment (geshuming) [reply src/components/assessment/assessmentShape.ts]: Done

File: src/components/assessment/assessmentShape.ts (hunk 10 above)
Comment (geshuming) [reply src/components/assessment/assessmentShape.ts]: I added comments to clarify what expected and actual means. I won't be changing this because the type follows the JSON format coming from the backend.

File: src/components/workspace/side-content/AutograderCard.tsx (hunk 11 above)
Comment (geshuming) [reply src/components/workspace/side-content/AutograderCard.tsx]: Done. I think it should be better now?

File: src/components/academy/grading/GradingWorkspace.tsx [hunk 12]
```diff
@@ -80,6 +83,23 @@ class GradingWorkspace extends React.Component<GradingWorkspaceProps> {
    */
   public componentDidMount() {
     this.props.handleGradingFetch(this.props.submissionId);
+
+    if (this.props.grading) {
```
[the remark this answers is not in the document]
Comment (geshuming) [reply src/components/academy/grading/GradingWorkspace.tsx]: Done as requested

File: src/components/academy/grading/GradingWorkspace.tsx (hunk 7 above)
Comment (geshuming) [reply src/components/academy/grading/GradingWorkspace.tsx]: Done as requested

File: src/components/academy/grading/GradingWorkspace.tsx (hunk 8 above)
Comment (geshuming) [reply src/components/academy/grading/GradingWorkspace.tsx]: Done as requested

File: src/components/assessment/assessmentShape.ts (hunk 10 above)
Comment (rrtheonlyone) [reply src/components/assessment/assessmentShape.ts]: Okay this is fine

File: src/components/workspace/side-content/AutograderCard.tsx (hunk 11 above)
Comment (rrtheonlyone) [reply src/components/workspace/side-content/AutograderCard.tsx]: We can improve it further actually! However, lets refer to my latest comment for now!

**Comment by geshuming** [comment]
@rrtheonlyone ok, I updated the typings, so by default the values should be `''` or `[]` if the backend responds with null values.

File: src/components/workspace/side-content/Autograder.tsx [hunk 13]
```diff
@@ -1,28 +1,85 @@
...
+type State = {
+  showTestcases: boolean;
+  showResults: boolean;
+};
+
+class Autograder extends React.Component<AutograderProps, State> {
+  public constructor(props: AutograderProps) {
+    super(props);
+
+    const hasResults: boolean = this.props.autogradingResults.length > 0 ? true : false;
```
Comment (rrtheonlyone) [inline src/components/workspace/side-content/Autograder.tsx]: `const hasResults: boolean = this.props.autogradingResults.length > 0;`

File: src/components/workspace/side-content/AutograderCard.tsx [hunk 14]
```diff
@@ -37,55 +13,39 @@ type AutograderCardProps = {
...
-        return <CanvasOutput />;
-      } else {
-        return stringify(value);
-      }
-    };
+    let gradingStatus: string = '';
+
+    if (this.props.testcase.result) {
+      gradingStatus =
+        stringify(this.props.testcase.result) === this.props.testcase.answer
```
Comment (rrtheonlyone) [inline src/components/workspace/side-content/AutograderCard.tsx]: remove ternary here also! The logic is not obvious

File: src/sagas/index.ts [hunk 15]
```diff
@@ -132,9 +132,9 @@ function* workspaceSaga(): SagaIterator {
     const location = (action as actionTypes.IAction).payload.workspaceLocation;
     const index = (action as actionTypes.IAction).payload.testcaseId;
     const code: string = yield select((state: IState) => {
-      const prepend = (state.workspaces[location] as IWorkspaceState).editorPrepend!;
+      const prepend = (state.workspaces[location] as IWorkspaceState).editorPrepend;
       const value = (state.workspaces[location] as IWorkspaceState).editorValue!;
-      const postpend = (state.workspaces[location] as IWorkspaceState).editorPostpend!;
+      const postpend = (state.workspaces[location] as IWorkspaceState).editorPostpend;
       const testcase = (state.workspaces[location] as IWorkspaceState).editorTestcases[index]
```
Comment (rrtheonlyone) [inline src/sagas/index.ts]: Can program be `null`?

File: src/sagas/index.ts [hunk 16]
```diff
@@ -438,6 +438,8 @@ function* evalTestCode(code: string, context: Context, location: WorkspaceLocati
       yield put(actions.evalTestcaseSuccess(result.value, location, index));
     } else {
       yield put(actions.evalInterpreterError(context.errors, location));
+      /* To indicate an error occured */
```
Comment (rrtheonlyone) [inline src/sagas/index.ts]: Remove redundant comment.

File: src/components/academy/grading/GradingWorkspace.tsx [hunk 17]
```diff
@@ -80,6 +83,25 @@ class GradingWorkspace extends React.Component<GradingWorkspaceProps> {
    */
   public componentDidMount() {
     this.props.handleGradingFetch(this.props.submissionId);
+
+    if (!this.props.grading) {
+      return;
+    }
+
+    const questionId =
+      this.props.questionId >= this.props.grading.length
```
[the remark this answers is not in the document]
Comment (geshuming) [reply src/components/academy/grading/GradingWorkspace.tsx]: Refactored

File: src/components/workspace/side-content/Autograder.tsx (hunk 13 above)
Comment (geshuming) [reply src/components/workspace/side-content/Autograder.tsx]: Refactored

File: src/components/academy/grading/GradingWorkspace.tsx [hunk 18]
```diff
@@ -80,6 +83,25 @@ class GradingWorkspace extends React.Component<GradingWorkspaceProps> {
...
+        ? this.props.grading.length - 1
+        : this.props.questionId;
+
+    const question: IQuestion = this.props.grading[questionId].question;
+    const answer: string =
+      question.type === QuestionTypes.programming
```
[the remark this answers is not in the document]
Comment (geshuming) [reply src/components/academy/grading/GradingWorkspace.tsx]: Refactored

File: src/components/workspace/side-content/AutograderCard.tsx (hunk 14 above)
Comment (geshuming) [reply src/components/workspace/side-content/AutograderCard.tsx]: Refactored

File: src/sagas/index.ts (hunk 15 above)
Comment (geshuming) [reply src/sagas/index.ts]: Removed

File: src/sagas/index.ts (hunk 16 above)
Comment (geshuming) [reply src/sagas/index.ts]: Removed

File: src/components/academy/grading/GradingWorkspace.tsx (hunk 17 above)
[the remark this answers is not in the document]
Comment (geshuming) [reply src/components/academy/grading/GradingWorkspace.tsx]: i think you meant `Math.min`, the idea is not to exceed the array index haha

[the remark this answers is not in the document]
Comment (geshuming) [reply src/components/academy/grading/GradingWorkspace.tsx]: actually, i dont even know why we have this guard, the component should redirect to 404 if an assessment is not found/question is not found

[the remark this answers is not in the document]
Comment (geshuming) [reply src/components/academy/grading/GradingWorkspace.tsx]: [I've logged an issue for this. I won't resolve it in this PR](https://github.com/source-academy/cadet-frontend/issues/623)

[the remark this answers is not in the document]
Comment (openorclose) [reply src/components/academy/grading/GradingWorkspace.tsx]: Whoops yeah i meant Math.min

File: src/components/workspace/side-content/AutograderCard.tsx [hunk 19]
```diff
@@ -37,55 +13,40 @@ type AutograderCardProps = {
...
   public render() {
-    const renderResult = (value: any) => {
-      /** A class which is the output of the show() function */
-      const ShapeDrawn = (window as any).ShapeDrawn;
-      if (typeof ShapeDrawn !== 'undefined' && value instanceof ShapeDrawn) {
-        return <CanvasOutput />;
+    let gradingStatus: string = '';
+
+    if (this.props.testcase.result) {
+      if (stringify(this.props.testcase.result) === this.props.testcase.answer) {
```
[the remark this answers is not in the document]
Comment (geshuming) [reply src/components/workspace/side-content/AutograderCard.tsx]: Done. Thanks for the lengthy review.
